src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,258 @@
+
+            var oldPower = MeasurementMode;
+
+            // Stop the magetometer
+            MeasurementMode = MeasurementMode.PowerDown;
+            // Enter the magnetometer Fuse mode to read the calibration data
+            MeasurementMode = MeasurementMode.FuseRomAccess;
+            // Read the data
+            ReadByteArray(Register.ASAX, rawData);
+            calib.X = (float)((rawData[0] - 128) / 256.0 + 1.0);
+            calib.Y = (float)((rawData[1] - 128) / 256.0 + 1.0);
+            calib.Z = (float)((rawData[2] - 128) / 256.0 + 1.0);
+            MeasurementMode = MeasurementMode.PowerDown;
+            MeasurementMode = oldPower;
+
+            // Now calculate the bias
+            // Store old mode to restore after
+            var oldMode = MeasurementMode;
+            Vector3 minbias = new Vector3();
+            Vector3 maxbias = new Vector3();
+            int numberMeasurements = 1500;
+
+            // Setup the 100Hz continous mode
+            MeasurementMode = MeasurementMode.ContinousMeasurement100Hz;
+            for (int reading = 0; reading < numberMeasurements; reading++)
+            {
+                var bias = ReadMagnetometer(true);
+                minbias.X = bias.X < minbias.X ? bias.X : minbias.X;
+                minbias.Y = bias.Y < minbias.Y ? bias.Y : minbias.Y;
+                minbias.Z = bias.Z < minbias.Z ? bias.Z : minbias.Z;
+                maxbias.X = bias.X > maxbias.X ? bias.X : maxbias.X;
+                maxbias.Y = bias.Y > maxbias.Y ? bias.Y : maxbias.Y;
+                maxbias.Z = bias.Z > maxbias.Z ? bias.Z : maxbias.Z;
+                Thread.Sleep(12);
+            }
+            // Store the bias
+            _magnetometerBias = ((maxbias + minbias) / 2) + calib;
+
+            return calib;
+        }
+
+        /// <summary>
+        /// True if there is a data to read
+        /// </summary>
+        public bool DataReady => (ReadByte(Register.ST1) & 0x01) == 0x01;
+
+        /// <summary>
+        /// Check if the version is the correct one (0x48)
+        /// </summary>
+        /// <returns>Returns true if the version match</returns>
+        public bool CheckVersion()
+        {
+            return ReadByte(Register.WIA) == 0x48;
+        }
+
+        /// <summary>
+        /// Read the magnetomer and can wait for new data to be present
+        ///         +X
+        ///  \  |  /
+        ///   \ | /
+        ///    \|/
+        ///    /|\
+        ///   / | \
+        ///  /  |  \
+        ///    +Z   +Y
+        /// </summary>
+        /// <param name="waitForData">true to wait for new data</param>
+        /// <returns>The data from the magnetometer</returns>
+        public Vector3 ReadMagnetometer(bool waitForData = false)
+        {
+            Span<byte> rawData = stackalloc byte[6];
+            // Wait for a data to be present
+            if (waitForData)
+            {
+                while (!DataReady)
```

Comment:
I think DataReady should become: WaitForDataReady/WaitForDataReadyAsync and used in here. Also presumably you can use that for calibration instead of delay

Reply:
I'm just using the Thread.Sleed in the calibration to avoid having access all the time to the register.
Will rename to WaitForDataReady